File: fabric-sdk-go/grant/macaroon_test.go
package grant

import (
	"bytes"
	"crypto/ed25519"
	cryptorand "crypto/rand"
	"errors"
	"strings"
	"testing"
	"time"
)

func mustRootKey(t *testing.T) []byte {
	t.Helper()
	k := make([]byte, 32)
	if _, err := cryptorand.Read(k); err != nil {
		t.Fatal(err)
	}
	return k
}

func sampleSpec(t *testing.T, rootKey []byte) MintSpec {
	t.Helper()
	pub, _, err := ed25519.GenerateKey(cryptorand.Reader)
	if err != nil {
		t.Fatal(err)
	}
	return MintSpec{
		RootKey:  rootKey,
		Location: "https://hub.bhai.example",
		Identifier: Identifier{
			GrantID:           "01JFXAMPLETESTGRANT00001",
			IssuedAt:          time.Date(2026, 5, 16, 12, 0, 0, 0, time.UTC),
			IssuedByPrincipal: "01JFXAMPLEHUMAN0000000001",
			IssuedByKeypair:   pub,
			Scope: Scope{
				Primitive:  PrimitiveVault,
				Namespace:  "list",
				Operations: []string{"read", "write"},
			},
		},
		Caveats: []string{
			"time < 2026-06-15T18:00:00Z",
			"operation in [read, write]",
		},
	}
}

func TestMintAndVerifyRoundTrip(t *testing.T) {
	rk := mustRootKey(t)
	g, err := Mint(sampleSpec(t, rk))
	if err != nil {
		t.Fatalf("Mint: %v", err)
	}
	if len(g.Macaroon) == 0 {
		t.Fatal("Macaroon bytes empty")
	}
	if g.GrantID == "" {
		t.Fatal("GrantID missing")
	}

	if err := VerifySignature(g.Macaroon, rk, AlwaysSatisfied); err != nil {
		t.Fatalf("VerifySignature: %v", err)
	}
}

func TestMintAndParsePreservesIdentifier(t *testing.T) {
	rk := mustRootKey(t)
	spec := sampleSpec(t, rk)
	g, err := Mint(spec)
	if err != nil {
		t.Fatal(err)
	}
	parsed, err := Parse(g.Macaroon)
	if err != nil {
		t.Fatalf("Parse: %v", err)
	}
	if parsed.Identifier.GrantID != spec.Identifier.GrantID {
		t.Errorf("GrantID: got %q, want %q", parsed.Identifier.GrantID, spec.Identifier.GrantID)
	}
	if parsed.Identifier.Scope.Primitive != spec.Identifier.Scope.Primitive {
		t.Errorf("Primitive: got %q, want %q", parsed.Identifier.Scope.Primitive, spec.Identifier.Scope.Primitive)
	}
	if !bytes.Equal(parsed.Identifier.IssuedByKeypair, spec.Identifier.IssuedByKeypair) {
		t.Error("IssuedByKeypair mismatch")
	}
	if len(parsed.Caveats) != len(spec.Caveats) {
		t.Fatalf("caveat count: got %d, want %d", len(parsed.Caveats), len(spec.Caveats))
	}
	for i, want := range spec.Caveats {
		if parsed.Caveats[i] != want {
			t.Errorf("caveat[%d]: got %q, want %q", i, parsed.Caveats[i], want)
		}
	}
}

func TestVerifyWrongKeyFails(t *testing.T) {
	rk1 := mustRootKey(t)
	rk2 := mustRootKey(t)
	g, err := Mint(sampleSpec(t, rk1))
	if err != nil {
		t.Fatal(err)
	}
	err = VerifySignature(g.Macaroon, rk2, AlwaysSatisfied)
	if err == nil {
		t.Fatal("VerifySignature with wrong key should fail")
	}
	if !errors.Is(err, ErrSignatureInvalid) {
		t.Errorf("errors.Is(err, ErrSignatureInvalid) should be true; got %v", err)
	}
}

func TestVerifyTamperedCaveatFails(t *testing.T) {
	rk := mustRootKey(t)
	g, err := Mint(sampleSpec(t, rk))
	if err != nil {
		t.Fatal(err)
	}
	// Flip a byte mid-payload. Locating exactly which byte is fiddly across
	// libmacaroon's binary encoding, so just touch the middle of the buffer
	// and assert the verify rejects it.
	g.Macaroon[len(g.Macaroon)/2] ^= 0x01
	if err := VerifySignature(g.Macaroon, rk, AlwaysSatisfied); err == nil {
		t.Fatal("VerifySignature on tampered macaroon should fail")
	}
}

func TestCheckFuncCanReject(t *testing.T) {
	rk := mustRootKey(t)
	g, err := Mint(sampleSpec(t, rk))
	if err != nil {
		t.Fatal(err)
	}
	rejectAll := func(c string) error { return errors.New("rejected: " + c) }
	err = VerifySignature(g.Macaroon, rk, rejectAll)
	if err == nil {
		t.Fatal("VerifySignature with rejecting CheckFunc should fail")
	}
	if !strings.Contains(err.Error(), "rejected:") {
		t.Errorf("expected rejection in error chain, got %v", err)
	}
}

File: fabric-sdk-js/test/macaroon.test.js
import { describe, it } from 'node:test';
import assert from 'node:assert/strict';
import {
  mint,
  parse,
  verifySignature,
  alwaysSatisfied,
  SignatureInvalidError,
} from '../src/grant/macaroon.js';
import { randomBytes } from '../src/crypto.js';

function sampleSpec(rootKey) {
  return {
    rootKey,
    location: 'https://hub.bhai.example',
    identifier: {
      grant_id: '01JFXAMPLETESTGRANT00001',
      issued_at: '2026-05-16T12:00:00Z',
      issued_by_principal: '01JFXAMPLEHUMAN0000000001',
      issued_by_keypair: randomBytes(32),
      scope: {
        primitive: 'vault',
        namespace: 'list',
        operations: ['read', 'write'],
      },
    },
    caveats: ['time < 2026-06-15T18:00:00Z', 'operation in [read, write]'],
  };
}

describe('Macaroon mint + verify', () => {
  it('round-trips signature', () => {
    const rk = randomBytes(32);
    const g = mint(sampleSpec(rk));
    assert.ok(g.macaroon.length > 0);
    assert.equal(g.grantId, '01JFXAMPLETESTGRANT00001');
    verifySignature(g.macaroon, rk, alwaysSatisfied);
  });

  it('parse preserves identifier and caveats', () => {
    const rk = randomBytes(32);
    const spec = sampleSpec(rk);
    const g = mint(spec);
    const parsed = parse(g.macaroon);
    assert.equal(parsed.identifier.grant_id, spec.identifier.grant_id);
    assert.equal(parsed.identifier.scope.primitive, 'vault');
    assert.deepEqual(parsed.identifier.issued_by_keypair, spec.identifier.issued_by_keypair);
    assert.equal(parsed.caveats.length, spec.caveats.length);
    for (let i = 0; i < spec.caveats.length; i++) {
      assert.equal(parsed.caveats[i], spec.caveats[i]);
    }
  });

  it('rejects wrong root key', () => {
    const rk1 = randomBytes(32);
    const rk2 = randomBytes(32);
    const g = mint(sampleSpec(rk1));
    assert.throws(() => verifySignature(g.macaroon, rk2, alwaysSatisfied), SignatureInvalidError);
  });

  it('rejects tampered macaroon', () => {
    const rk = randomBytes(32);
    const g = mint(sampleSpec(rk));
    g.macaroon[Math.floor(g.macaroon.length / 2)] ^= 0x01;
    assert.throws(() => verifySignature(g.macaroon, rk, alwaysSatisfied));
  });

  it('check function can reject', () => {
    const rk = randomBytes(32);
    const g = mint(sampleSpec(rk));
    assert.throws(
      () =>
        verifySignature(g.macaroon, rk, (c) => {
          throw new Error('rejected: ' + c);
        }),
      (err) => {
        assert.ok(err instanceof SignatureInvalidError);
        assert.match(String(err.cause?.message ?? ''), /rejected:/);
        return true;
      },
    );
  });
});
